src/lib/protocols/mdns/Publisher.cpp

```diff
@@ -0,0 +1,171 @@
+        SuccessOrExit(error = SetupHostname());
+    }
+    mIsPublishingProvisionedDevice = chip::DeviceLayer::ConfigurationMgr().IsFullyProvisioned();
+
+    if (mIsPublishingProvisionedDevice)
+    {
+        error = PublishProvisionedDevice(interface);
+    }
+    else
+    {
+        error = PublishUnprovisionedDevice(interface);
+    }
+exit:
+    return error;
+}
+
+CHIP_ERROR Publisher::SetupHostname()
+{
+    uint8_t mac[6];
+    char hostname[13]; // Hostname will be the hex representation of mac.
+    CHIP_ERROR error;
+
+    SuccessOrExit(error = chip::DeviceLayer::ConfigurationMgr().GetPrimaryWiFiMACAddress(mac));
+    for (size_t i = 0; i < sizeof(mac); i++)
+    {
+        snprintf(&hostname[i * 2], sizeof(hostname) - i * 2, "%02X", mac[i]);
+    }
+    SuccessOrExit(error = ChipMdnsSetHostname(hostname));
+
+exit:
+    return error;
+}
+
+CHIP_ERROR Publisher::PublishUnprovisionedDevice(chip::Inet::InterfaceId interface)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    MdnsService service;
+    uint16_t discriminator;
+    uint16_t vendorID;
+    uint16_t productID;
+    char discriminatorBuf[5];
+    char vendorProductBuf[10];
+    TextEntry entries[2] = {
+        { "D", nullptr, 0 },
+        { "VP", nullptr, 0 },
+    };
+
+    VerifyOrExit(mInitialized, error = CHIP_ERROR_INCORRECT_STATE);
+    ChipLogProgress(Discovery, "setup mdns service");
+    SuccessOrExit(error = chip::DeviceLayer::ConfigurationMgr().GetSetupDiscriminator(discriminator));
+    snprintf(service.mName, sizeof(service.mName), "%016" PRIx64, mUnprovisionedInstanceName);
+    strncpy(service.mType, "_chipc", sizeof(service.mType));
+    service.mProtocol = MdnsServiceProtocol::kMdnsProtocolUdp;
+    SuccessOrExit(error = chip::DeviceLayer::ConfigurationMgr().GetVendorId(vendorID));
+    SuccessOrExit(error = chip::DeviceLayer::ConfigurationMgr().GetProductId(productID));
+    snprintf(discriminatorBuf, sizeof(discriminatorBuf), "%04X", discriminator);
+    snprintf(vendorProductBuf, sizeof(vendorProductBuf), "%X+%X", vendorID, productID);
+    entries[0].mData       = reinterpret_cast<const uint8_t *>(discriminatorBuf);
+    entries[0].mDataSize   = strnlen(discriminatorBuf, sizeof(discriminatorBuf));
+    entries[1].mData       = reinterpret_cast<const uint8_t *>(vendorProductBuf);
+    entries[1].mDataSize   = strnlen(discriminatorBuf, sizeof(vendorProductBuf));
+    service.mTextEntryies  = entries;
+    service.mTextEntrySize = sizeof(entries) / sizeof(TextEntry);
+    service.mPort          = CHIP_PORT;
+
+    error = ChipMdnsPublishService(&service);
+
+exit:
+    return error;
+}
+
+CHIP_ERROR Publisher::PublishProvisionedDevice(chip::Inet::InterfaceId interface)
+{
+    uint64_t deviceId;
+    uint64_t fabricId;
+    MdnsService service;
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    SuccessOrExit(error = chip::DeviceLayer::ConfigurationMgr().GetFabricId(fabricId));
```

Comment:
This is again something that should likely not use singleton values, but have them passed down, in case the multiadmin conversations boil down to having > 1 device ID per physical node (which is not impossible).

Reply:
Also prefer to modify the platform api to iterate through the fabric/device ids.